protocol-units/settlement/mcr/contracts/src/MCR.sol

```diff
@@ -165,16 +261,19 @@ contract MCR {
     function submitBlockCommitmentForValidator(
         address validatorAddress, 
         BlockCommitment memory blockCommitment
-    ) external {
+    ) internal {
+
+        require(commitments[blockCommitment.height][validatorAddress].height == 0, "Validator has already committed to a block at this height");
 
-        require(commitments[blockCommitment.height][validatorAddress].height != 0, "Validator has already committed to a block at this height");
+        // note: do no uncomment the below, we want to allow this in case we have lagging validators
+        // require(blockCommitment.height > lastAcceptedBlockHeight, "Validator has committed to an already accepted block");
 
-        require(commitments[blockCommitment.height][validatorAddress].height < lastAcceptedBlockHeight + leadingBlockTolerance, "Validator has committed to a block too far ahead of the last accepted block");
+        require(blockCommitment.height < lastAcceptedBlockHeight + leadingBlockTolerance, "Validator has committed to a block too far ahead of the last accepted block");
 
         // assign the block height to the current epoch if it hasn't been assigned yet
-        if (!blockHeightEpochAssignments[blockCommitment.height] == 0) {
+        if (blockHeightEpochAssignments[blockCommitment.height] == 0) {
             // note: this is an intended race condition, but it is benign because of the tolerance
-            blockHeightEpochAssignments[blockCommitment.height] = getCurrentEpoch();
+            blockHeightEpochAssignments[blockCommitment.height] = getEpochByBlockTime();
```

Comment:
I did read the comment, but I just wonder if there are any unintended side effects from the race condition :thinking:

Reply:
This race lends to a simpler system than trying to have stake vote on this mapping. With the block height tolerance, you can ensure that it does not allow validators to run ahead and lock the validator set on arbitrarily many block heights; you can only be so far ahead of the work of the honest network. Without that tolerance check, that would be the primary attack vector.